### PR TITLE
Dropbox: Retrieves & uses refresh token to obtain new access token

### LINKED ISSUE #1267: Dropbox code needs to be updated to use refresh tokens
Dropbox no longer issues "long-lived tokens", only tokens that last about four hours: https://dropbox.tech/developers/migrating-app-permissions-and-access-tokens#updating-access-token-type  Users must click a link every four hours to re-authorize the app.

To regain the original user experience (what Dropbox calls "offline access"), the Dropbox code must be updated to use "Refresh tokens" to renew the access token.

### PR DESCRIPTION
**Pull request by DougReeder**
Fixes #1267

It may prove helpful to examine each commit separately.

The first commit adds Mocha tests for every scenario that jaribu tests, but re-arranged more logically. Some tests are pending, and some expectations are commented out (such as returning proper MIME types), because the current Dropbox backend is not compliant.

The second commit adds Content-Length & Last-Modified to Dropbox folder listings, correctly encodes Dropbox-API-Arg header and fixes some other bugs. The Dropbox-API-Arg header must be JSON-encoded, not URL-encoded: https://www.dropbox.com/developers/reference/json-encoding

The third commit refactors the function that calls either fetch or XHR into its own class. Authorize needs to call this as part of OAuth2 PKCE, and it was poor architecture for Dropbox and Google Drive to be using a function in wireclient.

The fourth commit refactors WireClient, GoogleDrive & Dropbox to have a common ancestor & Typescript interface. The common ancestor reduces duplication of code. The interface surfaces the complexity of interaction between the backends and other modules.

The fifth commit avoids the user needing to log in every four hours (what Dropbox calls "offline access" and OAuth2 calls the PKCE flow). OAuth2 PKCE is standardized, so I've placed as much code as possible in authorize.ts for future re-use.

It also sends the scopes required for Scoped apps, which Dropbox requires all new apps to be.

It also handles the 503 and 429 status codes and their Retry-After header that Dropbox sends.

The sixth commit removes the jaribu tests of new functionality that were superseded by better mocha tests.

### CODE REVIEW

File: src/requests.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        throw new Error("responseType 'document' is not currently supported using fetch");
+    }
+  }).then((processedBody) => {
+    syntheticXhr.response = processedBody;
+    if (!options.responseType || options.responseType === 'text') {
+      syntheticXhr.responseText = processedBody;
+    }
+    return syntheticXhr;
+  }).finally(() => {
+    clearTimeout(timeoutId);
```
Comment (DougReeder) [inline src/requests.ts:128]: This is the only significant difference between the functions here, and the old ones in wireclient.ts. Node isn't able to clean up obsolete timeouts.

**Review by raucao — COMMENTED** [review]
Fantastic work! :clap: 

I left a few comments for minor questions and nitpicks.

There's a breaking change for app developers with existing, older Dropbox apps, which we need to document (perhaps in a how-to-upgrade post on the forums and/or extended release notes):

Dropbox only introduced scopes fairly recently, and this PR requires the Dropbox app/client id to be configured with the new scopes. Otherwise authorization will fail:

Screenshot from 2022-10-20 11-56-53

The developer needs to go to https://www.dropbox.com/developers/apps, where there is already a message waiting for apps not using scopes to migrate the permissions:

Screenshot from 2022-10-20 12-13-09

On the permissions/scopes screen, one should deselect the permissions for contacts and file requests, then hit the migrate button:

Screenshot from 2022-10-20 12-14-38

After doing this, I was able to test these changes successfully with the RS Widget (which will now fail to authorize Dropbox until the widget is updated to use this rs.js changeset as well).

File: src/dropbox.ts [hunk 2]
```diff
@@ -476,28 +490,25 @@ class Dropbox {
...
+  async 'delete' (path, options) {
     if (!this.connected) {
       throw new Error("not connected (path: " + path + ")");
     }
 
     // check if file has changed and return 412
     const savedRev = this._revCache.get(path);
-    if (options && options.ifMatch && savedRev && (options.ifMatch !== savedRev)) {
-      return Promise.resolve({ statusCode: 412, revision: savedRev });
+    if (options?.ifMatch && savedRev && (options.ifMatch !== savedRev)) {
```
[the remark this answers is not in the document]
Comment (DougReeder) [reply src/dropbox.ts:559]: options?.ifMatch will evaluate as undefined if options is anything but an object: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

It certainly wouldn't hurt to define a default value of {} for options.

[the remark this answers is not in the document]
Comment (raucao) [reply src/dropbox.ts:559]: Ah, you're right. I had only tried to evaluate it in the console and that throws an error, but using it in a function works just fine.

File: src/dropbox.ts [hunk 3]
```diff
@@ -255,7 +277,7 @@ class Dropbox {
     };
 
     const loadNext = (cursor) => {
-      const continueURL = 'https://api.dropboxapi.com/2/files/list_folder/continue';
+      const continueURL = CONTINUE_URL;
```
[the remark this answers is not in the document]
Comment (DougReeder) [reply src/dropbox.ts:337]: That's a good idea.

File: src/dropbox.ts [hunk 4]
```diff
@@ -416,14 +416,16 @@ class Dropbox {
...
+   * @param {string} contentType - MIME type of body
    * @param {Object} options
    * @param {string} options.ifNoneMatch - When *, only create or update the file if it doesn't yet exist
    * @param {string} options.ifMatch - Only saves if this matches current revision
    * @returns {Promise} Resolves with an object containing the status code,
    *                    content-type and revision
    * @protected
    */
-  async put (path, body, contentType, options) {
+  async put (path: string, body, contentType: string, options: { ifMatch?: string; ifNoneMatch?: string } = {}) {
```
[the remark this answers is not in the document]
Comment (DougReeder) [reply src/dropbox.ts]: Unfortunately, there's a check of `body.length` below, which TypeScript will not allow, as several of the constituent types don't have that property. :-(

File: doc/getting-started/dropbox-and-google-drive.rst [hunk 5]
```diff
@@ -37,10 +37,15 @@ Dropbox
...
+    * account_info.read
+    * files.metadata.read
+    * files.metadata.write
+    * files.content.read
+    * files.content.write
 * You need to set one or more OAuth2 redirect URIs for all routes a user can
-  connect from, for example ``http://localhost:8000/`` for an app you are
-  developing locally.
+  connect from, for example ``http://localhost:8000`` for an app you are
+  developing locally.  If the path is '/', remoteStorage drops it.
```
[the remark this answers is not in the document]
Comment (DougReeder) [reply doc/getting-started/dropbox-and-google-drive.rst:48]: Yes, `authorize()` special-cases that, for some reason:
```

**Comment by DougReeder** [comment]
I believe the `sharing.write` permission isn't needed, just `account_info.read files.content.read files.content.write files.metadata.read files.metadata.write`

**Pull request merged**